add session_factory_from_settings and includeme

silenius opened the pull request:
Add session_factory_from_settings and includeme

silenius committed:
add session_factory_from_settings and includeme

silenius committed:
add myself to contributors, as requested

silenius committed:
update supported versions

silenius committed:
update supported versions

silenius committed:
fix typo

mmerickel commented on pyramid_nacl_session/utils.py:
```diff
@@ -0,0 +1,41 @@
+                    if v in serializers:
+                        v = serializers[v]()
+                elif option_name == 'secret':
+                    v = unhexlify(v)
```
I sort of think the secret should just be treated as text. At least I haven't seen precedent in other libraries for requiring it to be a hexlify'd string.

silenius replied:
secret must be a 32-byte string

silenius committed:
address comments

silenius committed:
secret should be a bytes string

stevepiercy commented on .travis.yml:
```diff
@@ -4,16 +4,16 @@ sudo: false
+        - python: 3.6
+          env: TOXENV=py36
+        - python: 3.7
+          env: TOXENV=py37
```
Please add Python 3.8.

stevepiercy commented on docs/usage.rst:
```diff
@@ -1,34 +1,38 @@
+Install using setuptools, e.g. (within a virtualenv)::
 
-.. code-block:: python
+  $ pip install pyramid_nacl_session
```
I think that installation of this package is seldom done by `pip install pyramid_nacl_session`, but usually by requiring it in `setup.py` or other methods. I think that we would better serve users by being vague about how to install this package ("Install pyramid_nacl_session via your preferred method.") or remove this section altogether.

However if you wish to preserve this section, then it should be:

```
Install using pip.

.. code-block:: bash

    pip install pyramid_nacl_session
```

stevepiercy commented on docs/usage.rst:
```diff
@@ -1,34 +1,38 @@
 
-Use the generated secret to configure a session factory:
+   config = Configurator(.....)
+   config.include('pyramid_nacl_session')
```
Five dots is not syntactically correct and looks weird. An actual example would be fine, and better to use a context manager with `with`.

```
    with Configurator(settings=settings) as config:
        config.include('pyramid_nacl_session')
        # ... more config.includes
        config.scan()
```

stevepiercy commented on docs/usage.rst:
```diff
@@ -1,34 +1,38 @@
-.. code-block:: python
+Alternately, instead of using the Configurator's ``include`` method, you can
+activate :mod:`pyramid_nacl_session` by changing your application's ``.ini`` 
+file, use the following line:
```
```suggestion
file, using the following line:
```

stevepiercy commented on docs/usage.rst:
``````````````````````diff
@@ -1,34 +1,38 @@
+-----
 
+Generate a secret key
+`````````````````````
``````````````````````
Please use heading underlines consistent with the levels in our [Style Guide](https://docs.pylonsproject.org/projects/docs-style-guide/#headings) and adopted from [Python](https://devguide.python.org/documenting/#sections). In this case, `^`.

stevepiercy commented on docs/usage.rst:
```````````````````diff
@@ -42,21 +46,57 @@ Cut-and-paste into your config file:
-Parse the secret from the configuration, de-hexlify it, and use it to
-construct and register a session factory:
+Session management
+``````````````````
```````````````````
Use `^` underline.

stevepiercy commented on docs/usage.rst:
```````````````````diff
@@ -42,21 +46,57 @@ Cut-and-paste into your config file:
-construct and register a session factory:
+Session management
+``````````````````
+If you have included pyramid_nacl_session in your Pyramid project's configuration
```````````````````
```suggestion
If you have included :mod:`pyramid_nacl_session` in your Pyramid project's configuration
```

stevepiercy commented on docs/usage.rst:
```diff
@@ -1,34 +1,38 @@
+-----
 
-   SECRET = generate_secret(as_hex=False)
+Once :mod:`pyramid_nacl_session` is installed, you typically use the 
```
When making changes to the docs, [please use one sentence per line](https://docs.pylonsproject.org/projects/docs-style-guide/#use-of-whitespace).

stevepiercy commented on docs/usage.rst:
```diff
@@ -42,21 +46,57 @@ Cut-and-paste into your config file:
-   from pyramid.session import BaseCookieSessionFactory
-   from pyramid_nacl_session import EncryptedCookieSessionFactory
+   [app:myapp]
+   .. other settings ..
```
```suggestion
   # other settings
```

stevepiercy commented on docs/usage.rst:
```diff
@@ -42,21 +46,57 @@ Cut-and-paste into your config file:
+       session_factory = session_factory_from_settings(settings)
+       config = Configurator(root_factory=get_root, settings=settings)
+       config.set_session_factory(session_factory)
+       # ... other configuration stuff...
```
Please see above example suggestion that uses a context manager and adapt accordingly.

stevepiercy commented on pyramid_nacl_session/utils.py:
```diff
@@ -0,0 +1,42 @@
+
+def session_factory_from_settings(settings):
+    """ Return a Pyramid session factory using PyNaCl session settings
+    supplied from a Paste configuration file"""
```
Nitpick: I prefer multi-line docstrings to have `"""` on their own separate lines. Nonetheless, docstrings must follow English punctuation rules.

```
    """
    Return a Pyramid session factory using PyNaCl session settings
    supplied from a Paste configuration file.
    """
```

stevepiercy commented on pyramid_nacl_session/utils.py:
```diff
@@ -0,0 +1,42 @@
+        'pickle': PickleSerializer
+    }
+
+    # Pull out any config args meant for PyNaCl session. if there are any
```
```suggestion
    # Pull out any config args meant for PyNaCl session, if there are any.
```

stevepiercy commented on tox.ini:
```diff
@@ -1,18 +1,18 @@
 envlist =
 #   py26,py27,py33,py34,py35,pypy,pypy3,
-    py26,py27,py33,py34,py35,pypy
+    py27,py34,py35,py36,py37,pypy
```
Add py38 here, too.

stevepiercy commented on tox.ini:
```diff
@@ -1,18 +1,18 @@
     py34: python3.4
     py35: python3.5
+    py36: python3.6
+    py37: python3.7
```
...and here.

silenius committed:
add Python 3.8

silenius committed:
make pylint happy

silenius committed:
fix comments, remove unused func

silenius committed:
adress reviewer comments

silenius committed:
fix typo

stevepiercy commented on docs/usage.rst:
```diff
@@ -50,53 +40,39 @@ Cut-and-paste into your config file:
+      with Configurator(settings=settings) as config:
+         config.set_session_factory(session_factory)
+         # other configuration stuff
+         return config.make_wsgi_app()
```
All the above lines 65-78 are not correctly indented. Please use 4 spaces for reStructuredText and Python code samples.

Also please check other indentation earlier in this file to ensure it is 4 spaces.

Note, PyCharm is notorious for inheriting the first indentation as its default for the document which will cause Python code samples to be incorrectly spaced. Keeping all indents to 4 spaces avoids that issue. Other editors might exhibit the same behavior.

silenius replied:
> Other editors might exhibit the same behavior.

indeed, same in Neovim :smiling_imp:

stevepiercy commented on pyramid_nacl_session/session.py:
```diff
@@ -1,22 +1,26 @@
+        reissue_time=0,
+        set_on_exception=True,
+        serializer=None,
 ):
```
Side note, if you use [Black](https://github.com/psf/black) for code formatting, I would not be opposed. We use it in Pyramid now.

silenius replied:
:heavy_check_mark:

stevepiercy commented on tox.ini:
```diff
@@ -1,7 +1,7 @@
 envlist =
 #   py26,py27,py33,py34,py35,pypy,pypy3,
-    py27,py34,py35,py36,py37,pypy
+    py27,py34,py35,py36,py37,py38,pypy
```
I missed a couple things on the first pass.

[Python 3.4 is now EOLed](https://www.python.org/dev/peps/pep-0429/#release-schedule). Suggest its removal as well, both here and below.

Also the commented line 3 should be dropped.

stevepiercy commented on .travis.yml:
```diff
@@ -14,6 +14,8 @@ matrix:
           env: TOXENV=py36
```
I missed this on my first pass. [Python 3.4 is now EOLed](https://www.python.org/dev/peps/pep-0429/#release-schedule). Suggest its removal as well, both here and below.

stevepiercy commented on setup.py:
```diff
@@ -11,7 +15,7 @@
 
 setup(name='pyramid_nacl_session',
-      version='0.4.dev0',
+      version=__version__,
```
Please update the trove classifiers below to support Python 3.5, 3.6, 3.7, 3.8.

stevepiercy reviewed (changes requested):
All your changes are good. I missed a couple of things on my first pass. One more round, and then let's get a maintainer to review. Thank you!

silenius committed:
fix indentation, remove Python 3.4

silenius committed:
format with black

stevepiercy reviewed (approved):
Thank you for your changes!

Next we need to figure out why Travis fails. I have no clue how to fix it.

Alternatively we could switch to GitHub Actions.

stevepiercy commented:
Er, I meant, I have no idea why the pypy build fails.

We need more test coverage for the changes, and that should take care of that one failure.

silenius commented:
I'll add some tests

silenius committed:
add some basic tests

silenius commented:
I've added some tests, (`utils.py` has now a coverage of 100%), but it looks like there is an issue with the PyPy instance, and some coverage issue for `__init__.py` ..

silenius committed:
cosmetic

stevepiercy commented:
I have no idea how to resolve those remaining issues. Maybe a maintainer has an idea?

silenius committed:
unhexlify secret

The bin/print_secret tool returns a 32-byte string encoded in
hexadecimal (suitable for copy/paste in some .ini file). We must
unhexlify the secret as the returned bin/print_secret string is twice as
long as the length of data.

silenius commented:
@mmerickel : I've readded the `unhexlify` in 44c2a6c otherwise we cannot use the string returned by `bin/print_secret` (because of https://github.com/Pylons/pyramid_nacl_session/blob/master/pyramid_nacl_session/serializer.py#L25-L28) ..

silenius commented:
@tseaver can you have a look once you have a little time? :relaxed:

tseaver commented on CHANGES.rst:
```diff
@@ -4,7 +4,9 @@ Changelog
-- TBD
+- Add session_factory_from_settings
+
+- Add includeme
```
Please note the removed / added support for Python version here as well.  I'd actually prefer to review those changes in a separate PR, but I understand how you go here.

tseaver commented on docs/index.rst:
```diff
@@ -1,20 +1,18 @@
-This Pyramid add-on defines an encrypting, pickle-based cookie serializer,
-using ``PyNaCl`` [1] to generate the symmetric encryption for the cookie state.
-
+This Pyramid add-on defines an encrypting, pickle-based cookie serializer, using ``PyNaCl`` [1] to generate the symmetric encryption for the cookie state.
```
Please wrap at 80 characters.

silenius replied:
ah, I thought documentation should not adhere to PEP8 maximum line length (https://docs.pylonsproject.org/projects/docs-style-guide/#use-of-whitespace)

tseaver commented on docs/usage.rst:
```diff
@@ -1,62 +1,78 @@
+-----
 
-First, you need to arrange to have a 32-byte random secret available:
+Once :mod:`pyramid_nacl_session` is installed, you typically use the ``config.include`` mechanism to include it into your Pyramid project's configuration.
```
Here and below, please wrap at 80 characters.

stevepiercy replied:
This request for change is inconsistent with https://docs.pylonsproject.org/projects/docs-style-guide/#use-of-whitespace

If we cannot agree on a standard as an organization, then we should revise this section in the guide, giving a consistent message to contributors.

tseaver commented on pyramid_nacl_session/__init__.py:
```diff
@@ -1,12 +1,12 @@
 from .session import EncryptedCookieSessionFactory
-_export(EncryptedCookieSessionFactory)
 from .scripts import generate_secret
-_export(generate_secret)
```
Rather than suppressing all `flake8` checks, please add the exported names to `__all__` to appease `flake8` (rather than the `_export` hack).

tseaver commented on pyramid_nacl_session/scripts.py:
```diff
@@ -27,22 +27,23 @@ def print_secret():  # pragma: NO COVER
     The secret is a 32-byte random string, encoded using
     :func:`binascii.hexlify`.
     """
+
```
Janitorial changes (e.g., blackening) should normally be done in a separate PR:  they make reviewing substantive changes much more difficult.

silenius replied:
Ok sorry, I'll think about it next time..!

tseaver commented on pyramid_nacl_session/serializer.py:
```diff
@@ -21,11 +21,13 @@ class EncryptedSerializer(object):
-                SecretBox.KEY_SIZE)
+                "Secret should be a random bytes string of length %d"
+                % SecretBox.KEY_SIZE
+            )
```
Rather than a gratuitous re-wrap, please use the more modern `.format()` syntax instead, e.g.:

```python
                "Secret should be a random bytes string of length {}".format(
                    SecretBox.KEY_SIZE
```

tseaver commented on pyramid_nacl_session/tests/test_session.py:
```diff
@@ -1,34 +1,115 @@
+from pyramid import testing
+
+
+class SomeClass:
```
Given that we are still supporting Python 2.7, use `object` for clarity:

```suggestion
class SomeClass(object):
```

tseaver reviewed (changes requested):
In addition to my inline comments:

- The PyPy test failure is due to PyPy 7.1.1 (the one which supplies the 2.7.13 API) shipping an older `cffi` library:

```bash
$ /opt/pypy-7.1.1-linux_x86_64-portable/bin/virtualenv-pypy /tmp/foobar
New pypy executable in /tmp/foobar/bin/pypy
Installing setuptools, pip, wheel...
done.
$ /tmp/foobar/bin/pip list
DEPRECATION: A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
Package    Version
---------- -------
cffi       1.12.3 
greenlet   0.4.13 
pip        19.3.1 
readline   6.2.4.1
setuptools 42.0.2 
wheel      0.33.6 
```

`PyNacl` needs `cffi >= 1.4.1`, but [doesn't enforce that on PyPy](https://github.com/pyca/pynacl/blob/4ff1d3d52bc298108f56b533ba3675670457f2ff/setup.py#L45-L53).

- The coverage failure is for the newly-added `includeme` function.

silenius committed:
add __all__ to appease flake8

silenius committed:
use object for clarity

silenius committed:
add a note about removed / added support for Python version

silenius committed:
80 characters wrap

silenius committed:
use the more modern .format() syntax

silenius committed:
add a test for includeme

silenius commented:
> The coverage failure is for the newly-added `includeme` function.

I just added a test for the newly-added `includeme` function.

stevepiercy commented:
@tseaver for the PyPy failure, what should we do? File an issue or PR against PyNacl? Merge this PR as is?

digitalresistor commented on pyramid_nacl_session/session.py:
```diff
@@ -1,15 +1,19 @@
+# -*- coding: utf-8 -*-
```
Please do not add this to Python files in this repository. The default is UTF-8 on Python 3 anyway.

digitalresistor commented on pyramid_nacl_session/session.py:
```diff
@@ -1,15 +1,19 @@
 from .serializer import EncryptedSerializer
 
+
+# pylint: disable=too-many-arguments,invalid-name
```
I don't believe we currently use pylint on this repository. Please do not add extra linter information.

digitalresistor commented on pyramid_nacl_session/session.py:
```diff
@@ -93,10 +97,7 @@ def EncryptedCookieSessionFactory(
-        secret,
-        serializer=serializer,
-    )
+    encrypted_serializer = EncryptedSerializer(secret, serializer=serializer,)
```
```suggestion
    encrypted_serializer = EncryptedSerializer(secret, serializer=serializer)
```

digitalresistor commented on pyramid_nacl_session/tests/test_session.py:
```diff
@@ -1,34 +1,135 @@
-from pyramid import testing
+# -*- coding: utf-8 -*-
```
Please do not add unnecessary changes to this PR.

digitalresistor commented on setup.py:
```diff
@@ -1,3 +1,7 @@
+# -*- coding: utf-8 -*-
+
+__version__ = '0.4.dev2'
```
Please do not move the version information into `__version__`, this is unnecessary and not a convention we use for Pylons Projects.

digitalresistor commented on pyramid_nacl_session/utils.py:
```diff
@@ -0,0 +1,42 @@
+    supplied from a Paste configuration file.
+    """
+
+    prefixes = ("session.", "pynacl.session.")
```
I am not a huge fan of using a `session.` prefix without adding support for this to other session libraries or potentially to Pyramid directly. It may cause confusion because its not supported for anything else.

silenius replied:
`pyramid_beaker` supports `session.` too

digitalresistor replied:
I'm a -0 on it, but I'll allow it.

digitalresistor commented on pyramid_nacl_session/__init__.py:
```diff
@@ -1,12 +1,15 @@
-    """
+# -*- coding: utf-8 -*-
 
+# flake8: noqa
```
Please remove this unnecessary line.

silenius committed:
changes requested by @bertjwregeer

digitalresistor commented on docs/index.rst:
```diff
@@ -1,20 +1,19 @@
 
 .. toctree::
-   :maxdepth: 2
+    :maxdepth: 2
```
I didn't notice this before, but the 3 spaces vs the 4 is intentional. In restructured text documents you match the `.. ` spacing not the usual 4 spaces in Python code.

stevepiercy replied:
We want 4 spaces for consistency. Editors will incorrectly indent Python code in `code-block` because they interpret the first indent as the file's default.

digitalresistor replied:
Oh, okay.

digitalresistor commented on docs/index.rst:
```diff
@@ -1,20 +1,19 @@
 ===========================
 
-This Pyramid add-on defines an encrypting, pickle-based cookie serializer,
+This Pyramid add-on defines an encrypting, pickle-based cookie serializer, 
```
Space at the end of a line is likely to get stripped by the next person to edit the file (I have it set up to do it automatically on files since in most cases it's a mistake), it is likely better to add this to the beginning of the next line.

digitalresistor commented on setup.py:
```diff
@@ -11,7 +11,7 @@
 
 setup(name='pyramid_nacl_session',
-      version='0.4.dev0',
+      version='0.4',
```
Don't drop the `.dev0`, the person who does the release will modify this value when the time comes.

digitalresistor commented:
Thank you @silenius! Just a couple of other comments and I will happily sign off on it. We can fix the pypy failure in a separate PR :-).

silenius committed:
don't change version in setup.py + fix whitespace

silenius commented:
> Thank you @silenius! Just a couple of other comments and I will happily sign off on it. We can fix the pypy failure in a separate PR :-).

Thank you for reviewing this :+1:

silenius committed:
add a reference to the pull request

silenius commented:
One last thing, should we document somewhere (`CHANGES.rst` ?) the namespace change from `yourapp.session_secret` to `session.` ?

silenius commented:
ping ? :smiley:

digitalresistor commented:
Thank you @silenius. Made an executive decision here.
